rest-service/manager_rest/rest/filters_utils.py

```diff
@@ -1,129 +1,138 @@
-    """Validate and parse a labels filter
-
-    :param labels_filter: One of <key>=<value>, <key>=[<value1>,<value2>,...],
-           <key>!=<value>, <key>!=[<value1>,<value2>,...]
-    :param sign: Either '=' or '!='
-    :return: The labels_filter, with its key and value(s) in lowercase and
-             stripped of whitespaces
-    """
-    try:
-        raw_label_key, raw_label_value = labels_filter.split(sign)
-    except ValueError:  # e.g. a=b=c
-        raise BadLabelsFilter(labels_filter)
-
-    label_key = raw_label_key.strip().lower()
-    label_values = _get_label_value(raw_label_value.strip().lower())
-    value_msg_prefix = (None if len(label_values) == 1 else
-                        'One of the filter values')
-
-    parsed_label_values_list = []
-    for value in label_values:
-        try:
-            parsed_value = value.strip()
-            validate_inputs({'filter key': label_key})
-            validate_inputs({'filter value': parsed_value},
-                            err_prefix=value_msg_prefix)
-        except manager_exceptions.BadParametersError as e:
-            err_msg = 'The filter rule {0} is invalid. '.format(labels_filter)
-            raise manager_exceptions.BadParametersError(err_msg + str(e))
-
-        parsed_label_values_list.append(parsed_value)
-
-    return label_key, parsed_label_values_list
-
-
-def _get_label_value(raw_label_value):
-    if raw_label_value.startswith('[') and raw_label_value.endswith(']'):
-        return raw_label_value.strip('[]').split(',')
-
-    return [raw_label_value]
+            raise BadFilterRule(filter_rule,
+                                f"Filter rule type must be one of "
+                                f"{', '.join(FILTER_RULE_TYPES)}")
+
+        value_msg_prefix = (None if len(filter_rule_values) == 1 else
+                            'One of the filter rule values')
+
+        parsed_values_list = []
+        for value in filter_rule_values:
+            try:
+                validate_inputs({'filter rule key': filter_rule_key})
+                validate_inputs({'filter rule value': value},
+                                err_prefix=value_msg_prefix)
+            except BadParametersError as e:
+                err_msg = f'The filter rule {filter_rule} is invalid. '
+                raise BadParametersError(err_msg + str(e))
+            parsed_values_list.append(value.lower())
+
+        parsed_filter_rules.append({'key': filter_rule_key.lower(),
+                                    'values': parsed_values_list,
+                                    'operator': filter_rule_operator,
+                                    'type': filter_rule_type})
+
+    return parsed_filter_rules
+
+
+def _assert_filter_rule_structure(filter_rule):
+    if not isinstance(filter_rule, dict):
+        raise BadFilterRule(filter_rule, 'The filter rule is not a dictionary')
+
+    if not (all(key in filter_rule for key in
+                ['key', 'values', 'operator', 'type'])):
```

Comment:
```suggestion
    if not all(key in filter_rule for key in
               ['key', 'values', 'operator', 'type']):
```

Reply:
what about additional keys in filter_rule?
we could also do `if filter_rule.keys() != {'key', 'values', 'operator', 'type'}:`

Reply:
Correct.